editor/src/components/canvas/controls/guideline-controls.tsx

```diff
@@ -153,3 +208,158 @@ function useGuideline<T = HTMLDivElement>(
   }, [innerCallback])
   return controlRef
 }
+
+function framesFromMetadata(
+  metadataMap: ElementInstanceMetadataMap,
+  paths: Array<ElementPath>,
+): Array<CanvasRectangle> {
+  return mapDropNulls(
+    (path) => MetadataUtils.findElementByElementPath(metadataMap, path)?.globalFrame ?? null,
+    paths,
+  )
+}
+
+interface CanvasSpan {
+  a: CanvasPoint
+  b: CanvasPoint
+}
+
+function canvasSpan(a: CanvasPoint, b: CanvasPoint): CanvasSpan {
+  return { a: a, b: b }
+}
+
+function guidelineToSpan(guideline: Guideline): CanvasSpan | null {
+  switch (guideline.type) {
+    case 'XAxisGuideline': {
+      const a = canvasPoint({ x: guideline.x, y: guideline.yBottom })
+      const b = canvasPoint({ x: guideline.x, y: guideline.yTop })
+      return canvasSpan(a, b)
+    }
+    case 'YAxisGuideline': {
+      const a = canvasPoint({ x: guideline.xLeft, y: guideline.y })
+      const b = canvasPoint({ x: guideline.xRight, y: guideline.y })
+      return canvasSpan(a, b)
+    }
+    case 'CornerGuideline':
+      return null
+    default:
+      return Utils.assertNever(guideline)
+  }
+}
+
+function guidelineEndpoints(guideline: Guideline): Array<CanvasPoint> {
+  switch (guideline.type) {
+    case 'XAxisGuideline': {
+      const a = canvasPoint({ x: guideline.x, y: guideline.yBottom })
+      const b = canvasPoint({ x: guideline.x, y: guideline.yTop })
+      return [a, b]
+    }
+    case 'YAxisGuideline': {
+      const a = canvasPoint({ x: guideline.xLeft, y: guideline.y })
+      const b = canvasPoint({ x: guideline.xRight, y: guideline.y })
+      return [a, b]
+    }
+    case 'CornerGuideline':
+      return []
+    default:
+      return Utils.assertNever(guideline)
+  }
+}
+
+function rectangleBoundingLines(rectangle: CanvasRectangle): CanvasSpan[] {
+  return [
+    canvasSpan(
+      canvasPoint({ x: rectangle.x, y: rectangle.y }),
+      canvasPoint({ x: rectangle.x, y: rectangle.y + rectangle.height }),
+    ),
+    canvasSpan(
+      canvasPoint({ x: rectangle.x, y: rectangle.y }),
+      canvasPoint({ x: rectangle.x + rectangle.width, y: rectangle.y }),
+    ),
+    canvasSpan(
+      canvasPoint({ x: rectangle.x + rectangle.width, y: rectangle.y }),
+      canvasPoint({ x: rectangle.x + rectangle.width, y: rectangle.y + rectangle.height }),
+    ),
+    canvasSpan(
+      canvasPoint({ x: rectangle.x, y: rectangle.y + rectangle.height }),
+      canvasPoint({ x: rectangle.x + rectangle.width, y: rectangle.y + rectangle.height }),
+    ),
+  ]
+}
+
+// https://stackoverflow.com/a/9997374
+function spansIntersect(a: CanvasSpan, b: CanvasSpan): boolean {
+  function ccw(p1: CanvasPoint, p2: CanvasPoint, p3: CanvasPoint): boolean {
+    return (p3.y - p1.y) * (p2.x - p1.x) <= (p2.y - p1.y) * (p3.x - p1.x)
+  }
+
+  return ccw(a.a, b.a, b.b) !== ccw(a.b, b.a, b.b) && ccw(a.a, a.b, b.a) !== ccw(a.a, a.b, b.b)
+}
+
+function spansIntersect2(a: CanvasSpan, b: CanvasSpan): boolean {
```

Comment:
Having a `spansIntersect` and a `spansIntersect2` and `spanIntersection` is confusing, what is the difference? Can we give them more meaningful names?

Reply:
while trying to explain the code in a comment, i actually found a better solution, so I pushed that instead (and the site that "inspired" the solution)